### ISSUE TITLE
NVIDIA nv0 training stalls beyond 60s on quantized layer/MoE model with batch normalization and custom float arithmetic

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#365]
Public `recipe.train().optimizer(adamw).lr(0.0001).seed(59985129387313).epochs(1).log(all).stop(0.8).f(6, 9).run()` on `backend:nv0` stalls before any epoch arithmetic. The model (`recipe.model().layer(4).silu().qi(4).k.s .layer(8) .moe(1, [layer(8), layer(8)]).ln().norm(batch).iq(3).xs .loss(rmse)`) lowers through `lower_project` (layer experts), `lower_moe` (`recipe.rs` ~5419), normalization emission, and quantized nodes `.qi()`/`.iq()` with `.f(6,9)` arithmetic; the stall reaches `compile_native_artifact` (`recipe.rs` 2828-2896, NVIDIA `ptx`/`ptxas` branch) at `compile_model`. The loader (`data/temporal/window_subfolders`) provides ~2304 rows and 1 feature; the small width-8 / 2-expert / single-epoch workload justifies sub-second execution, so the 60-second gap is the native-artifact compilation stall, not arithmetic or memory traffic.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `be16fa50d757948a`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "NVIDIA nv0 training stalls beyond 60s on quantized layer/MoE model with batch normalization and custom float arithmetic",
  "body": "Public `recipe.train().optimizer(adamw).lr(0.0001).seed(59985129387313).epochs(1).log(all).stop(0.8).f(6, 9).run()` on `backend:nv0` stalls before any epoch arithmetic. The model (`recipe.model().layer(4).silu().qi(4).k.s .layer(8) .moe(1, [layer(8), layer(8)]).ln().norm(batch).iq(3).xs .loss(rmse)`) lowers through `lower_project` (layer experts), `lower_moe` (`recipe.rs` ~5419), normalization emission, and quantized nodes `.qi()`/`.iq()` with `.f(6,9)` arithmetic; the stall reaches `compile_native_artifact` (`recipe.rs` 2828-2896, NVIDIA `ptx`/`ptxas` branch) at `compile_model`. The loader (`data/temporal/window_subfolders`) provides ~2304 rows and 1 feature; the small width-8 / 2-expert / single-epoch workload justifies sub-second execution, so the 60-second gap is the native-artifact compilation stall, not arithmetic or memory traffic.",
  "rationale": "Performance packet `kind=performance`, `stable:true`, measured 60s on `nv0`. Source inspection confirms earliest failure is `recipe.train().run()` stalling at `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` (`recipe.rs` 2828-2896) before any training arithmetic. The reproduction composition uses `.layer()` / `.moe(1, [layer(8), layer(8)])` with `.ln()` activation, `.norm(batch)`, `.qi(4).k.s` and `.iq(3).xs` quantization, and `.f(6,9)` arithmetic. This reaches `lower_project` and `lower_moe` (`recipe.rs` ~5419), a quantized layer/MoE lowering path that is not covered by existing open nv0 stall issues (271 residual only, 315 perc/MoE integer arithmetic, 363 residual/MoE custom float, 364 conv/MoE custom float). The dataset (`window_subfolders`, ~96 subfolders of ~24 rows each, 1 feature) and small model dimensions make the 60-second stall materially unjustified; expected runtime is sub-second. The loader, activation, normalization, and arithmetic settings are irrelevant to the earliest cause; they are additional evidence for the same `compile_native_artifact` terminal check on this distinct layer/MoE quantized path, requiring a new entry."
}
```

## Deterministic failure packet

kind=performance
id=be16fa50d757948a
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:11147 next:11148 composition:59985129387313
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-issue-machine/target/debug/recipe-trials/recipe-trial-70-11147/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/window_subfolders")
		.target("target")
		.broadcast();
	let model = recipe.model()
		.layer(4).silu().qi(4).k.s
		.layer(8)
		.moe(1, [layer(8), layer(8)]).ln().norm(batch).iq(3).xs
		.loss(rmse);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(59985129387313)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.f(6, 9)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Failed on Thursday, August 27, 2026 2:36 PM PDT

Commit: 8f386573e4b6979d7e4d03204602144888a2ebe6

Command:
```text
recipe test.rs
```

Reproduction script:
```rust
use recipe::*;
use std::time::{Duration, Instant};

fn main() {
	let bundle = format!("/tmp/recipe-issue-365-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/window_subfolders")
		.target("target")
		.broadcast();
	let model = recipe.model()
		.layer(4).silu().qi(4).k.s
		.layer(8)
		.moe(1, [layer(8), layer(8)]).ln().norm(batch).iq(3).xs
		.loss(rmse);
	let started = Instant::now();
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(59985129387313)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.f(6, 9)
		.save(&bundle)
		.run(&model, &data);
	let elapsed = started.elapsed();
	let final_loss = report.final_loss();
	println!("final_loss_finite={}", final_loss.is_finite());
	println!("elapsed_seconds={:.3}", elapsed.as_secs_f64());
	println!("within_60_seconds={}", elapsed < Duration::from_secs(60));
	assert!(final_loss.is_finite());
	assert!(elapsed < Duration::from_secs(60));
}
```

Full output:
```text
recipe: unknown command "test.rs"; expected `probe`

Recipe command line

Usage:
  recipe probe [OPTIONS]

The zero-argument probe uses the embedded theoretical seed contract, discovers
the current bare-metal machine, benchmarks every discovered device and link,
and writes an identity-keyed measured profile under the user's private cache.

Options:
  --contract PATH       Use another theoretical probe seed contract
  --profile PATH        Write/load this absolute measured-profile path
  --cuda-driver PATH    Exact CUDA Driver library candidate (repeatable)
  --hsa-runtime PATH    Exact ROCr/HSA runtime candidate (repeatable)
  --llvm-opt PATH       Exact LLVM IR verifier
  --llvm-llc PATH       Exact LLVM code generator
  --lld PATH            Exact ELF linker
  --ptxas PATH          Exact NVIDIA PTX assembler
  -h, --help             Show this help

```

Verdict: FAIL
Reason: The required public `recipe test.rs` path is absent because the `recipe` executable accepts only `probe`. The reproduction never reaches model compilation or the one-epoch nv0 run, so finite loss, completion within 60 seconds, and throughput proportionate to the stated workload remain unproven.
